### ISSUE TITLE
AI ML Study continued 

### ISSUE THREAD

**Issue by kurtzace** [kurtzace/diary-2025#2]
continued from https://github.com/kurtzace/diary-2024/issues/7 

current material 
### CNN 2

example cnn
```
from functools import partial

DefaultConv2D = partial(tf.keras.layers.Conv2D, kernel_size=3, padding="same",
                        activation="relu", kernel_initializer="he_normal")
model = tf.keras.Sequential([
    DefaultConv2D(filters=64, kernel_size=7, input_shape=[28, 28, 1]),
    tf.keras.layers.MaxPool2D(),
    DefaultConv2D(filters=128),
    DefaultConv2D(filters=128),
    tf.keras.layers.MaxPool2D(),
    DefaultConv2D(filters=256),
    DefaultConv2D(filters=256),
    tf.keras.layers.MaxPool2D(),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(units=128, activation="relu",
                          kernel_initializer="he_normal"),
    tf.keras.layers.Dropout(0.5),
    tf.keras.layers.Dense(units=64, activation="relu",
                          kernel_initializer="he_normal"),
    tf.keras.layers.Dropout(0.5),
    tf.keras.layers.Dense(units=10, activation="softmax")
])
```

- AlexNet also uses a competitive normalization step - after the ReLU step of c1..c3 layers, called local response normalization (LRN): the most strongly activated neurons inhibit other neurons located at the same position in neighboring feature maps.

```
tf.nn.local_response_normalization()
```

Keras’s data augmentation layers, (e.g., RandomCrop, RandomRotation, etc.). This forces the model to be more tolerant to variations in the position, orientation, and size of the objects in the pictures.

```
data_augmentation = tf.keras.Sequential([
    tf.keras.layers.RandomFlip(mode="horizontal", seed=42),
    tf.keras.layers.RandomRotation(factor=0.05, seed=42),
    tf.keras.layers.RandomContrast(factor=0.2, seed=42)
])
```

- GoogLeNet: inception module
![image](https://github.com/user-attachments/assets/b15d19d4-90c4-424f-8d4a-4f4ef5d54efa)
 a. first two layers divide the image’s height and width by 4
b. local norm 
c.half by max pool
d. nine inception modules
e. global average pooling layer outputs the mean of each feature map

- ResNet : CNN composed of 152 layers
skip connections (also called shortcut connections)
![image](https://github.com/user-attachments/assets/05e7a7d8-8991-43b5-b727-9f104f5b0417)

BN may be batch norm

- Xception 
replaces the inception modules with a special type of layer called a depthwise separable convolution layer (or separable convolution layer for short⁠
![image](https://github.com/user-attachments/assets/2a5ecea6-28d7-4752-a3b6-d835103022c6)

-  SENET **[Squeeze-and-Excitation Network (SENet)](https://homl.info/senet).⁠**
se block
![image](https://github.com/user-attachments/assets/b1522072-10fb-4be8-ae55-adcdf742dba0)

![image](https://github.com/user-attachments/assets/596df5e2-9ae9-47e7-ab4b-727ab8653316)

- MobileNets are streamlined models designed to be lightweight / fast

- [EfficientNet](https://homl.info/efficientnet)⁠ 
- https://medium.com/@julianaliu1997/keras-implementation-of-resnet-34-residual-networks-architecture-along-with-variants-resnet-50-c5314bd5c70f ResNet-50 and ResNeXt 
or best is 

`model = tf.keras.applications.ResNet50(weights="imagenet")`

or
`base_model = tf.keras.applications.xception.Xception(weights="imagenet",
                                                     include_top=False)`

```
images = load_sample_images()["images"]
images_resized = tf.keras.layers.Resizing(height=224, width=224,
                                          crop_to_aspect_ratio=True)(images)
inputs = tf.keras.applications.resnet50.preprocess_input(images_resized)
Y_proba = model.predict(inputs)
#output Y_proba is a matrix with one row per image and one column per class 
top_K = tf.keras.applications.resnet50.decode_predictions(Y_proba, top=3)
```

freeze weeights

```
base_model = tf.keras.applications.xception.Xception(weights="imagenet", include_top=False)
avg = tf.keras.layers.GlobalAveragePooling2D()(base_model.output)
output = tf.keras.layers.Dense(n_classes, activation="softmax")(avg)
model = tf.keras.Model(inputs=base_model.input, outputs=output)
for layer in base_model.layers: # or base_model.layers[56:]:
    layer.trainable = False
optimizer = tf.keras.optimizers.SGD(learning_rate=0.1, momentum=0.9)
model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer,
              metrics=["accuracy"])
history = model.fit(train_set, validation_data=valid_set, epochs=3)
```

The most common metric for this is the intersection over union (IoU): the area of overlap between the predicted bounding box and the target bounding box,

**Comment by kurtzace** [comment]
RNN into a sequence-to-sequence RNN. The advantage of this technique is that the loss will contain a term for the output of the RNN at each and every time step, not just for the output at the last time step.

**Comment by kurtzace** [comment]
# GenAI

## AWS bedrock
[pluralsight path](https://app.pluralsight.com/paths/skill/amazon-bedrock)

Policy: marketplace view subscriptions, unsubscribe, subscribe

Models: A121, Amazon, Anthropic, Cohere, Meta, Mistral AI, Stability - Model access - request

Playground: Chat/text or Image

Titan Text G1:  4096 resp length - for quick tasks

Titan Image Generator G1: Gen image, gen variations, remove object, replace object, replace background. Add neg prompt. Size of image, # of images. 

Chat or single prompt. 


Prompt management.

Knowledge base for private data

Agent for multistep workflow


prompt flows: workflows





### interesting word learnt
Kaizen - continious learnings
